### ISSUE TITLE
Automated fine-tuning pipeline

### ISSUE THREAD

**Issue by mattg-stack** [Glad-Labs/poindexter#49]
*Migrated from Gitea Glad-Labs/poindexter#220*

---

## Problem

Fine-tuning a model once is a learning exercise. Keeping it current as the content library grows and the brand voice evolves requires automation. Manual re-training doesn't scale.

## Proposed feature

An automated pipeline that periodically re-fine-tunes the poindexter-writer model as new high-quality content is published.

### How it would work

```
New post published + QA score >= 80
    ↓
Training data builder (extract instruction/completion pairs)
    ↓
Append to training dataset (versioned in DB or S3)
    ↓
Trigger condition met? (e.g., 10 new posts since last train)
    ↓
Fine-tuning job (QLoRA, runs on 5090 GPU)
    ↓
Evaluation (compare new model vs current on held-out test set)
    ↓
If improved → promote to production (swap Ollama model)
    ↓
Log results to brain_decisions table
```

#### 1. Training data builder
- Extracts published posts from DB
- Formats as instruction/completion pairs for fine-tuning
- Filters: only posts with QA score >= threshold
- Deduplicates and versions the dataset
- Stores in a structured format (JSONL)

#### 2. Fine-tuning orchestrator
- Wraps Unsloth/Axolotl fine-tuning in a managed process
- Configurable hyperparameters (stored in app_settings)
- GPU scheduling — only runs when the 5090 isn't doing content generation
- Exports to GGUF for Ollama consumption

#### 3. Evaluation harness
- Holds out 10-20% of training data as test set
- Runs both old and new model on test prompts
- Compares QA scores, content_validator pass rates, and style consistency
- Auto-promotes if new model beats old model on all metrics
- Rolls back if regression detected

#### 4. Model registry
- Tracks model versions, training data used, eval scores
- Stored in DB (new `model_versions` table)
- Supports rollback to any previous version
- Integrates with the model router

### Integration points

| System | Integration |
|--------|------------|
| Content pipeline | Triggers training data capture on publish |
| Brain daemon | Monitors model freshness, triggers re-training |
| Model router | Swaps active model when new version promoted |
| Cost guard | Tracks fine-tuning GPU hours as operational cost |
| Grafana | Dashboard panel for model version, last trained, eval scores |

### app_settings keys

```
fine_tune_min_posts = 10        # Min new posts before re-training
fine_tune_qa_threshold = 80     # Min QA score to include in training
fine_tune_eval_threshold = 0.05 # New model must beat old by 5%
fine_tune_base_model = glm-4-9b # Base model for fine-tuning
fine_tune_schedule = weekly     # Or "on_threshold" (trigger-based)
```

### Why automate?

- **Model drift**: As writing style evolves, the model should evolve too
- **Continuous improvement**: Every good post makes the next post better
- **Zero manual work**: Fits the autonomous pipeline philosophy
- **Consulting selling point**: "The system trains itself on your content"

### Dependencies

- Issue Glad-Labs/poindexter#219 (poindexter-writer model) — need the first manual fine-tune working
- Sufficient published posts (50+ for initial, ongoing for continuous)
- GPU scheduling (don't compete with content generation)

### Milestone

Backlog — this is post-M3 at earliest. Build the manual fine-tuning first (issue Glad-Labs/poindexter#219), prove it works, then automate.

**Comment by mattg-stack** [comment]
Triaged 2026-05-08 — needs your call. Now that Glad-Labs/glad-labs-stack#1131 (fine-tune-our-own-writer) is closed in favor of glm-4.7-5090 + DB-configurable model selection, this auto-FT pipeline is orphaned. `project_custom_model_training` keeps fine-tuning as a long-term vision, so killing this would close one of the loops you said you wanted. **Recommendation:** keep as P3 backlog under the long-term vision, but mark blocked-by:#82 (need the dataset extraction live first). Reply here or on Telegram to confirm.

**Comment by mattg-stack** [comment]
Issue #82 blocking
